tests/testthat/test-derive_expected_records.R

```diff
@@ -0,0 +1,43 @@
+## Test 1: `by_vars` are missing ----
+test_that("derive_expected_records Test 1: `by_vars` are missing", {
```

Comment:
The title does not match the code.
Could we have two tests, one with `by_vars` missing and one with `by_vars` not missing?

Reply:
how about the below four sets? 
1. missing values in `by_vars`
2. `by_vars` = `NULL`
3. visit variables are parameter independent
4. visit variables are parameter dependent

Reply:
Looks good.

Reply:
the second case `by_vars = NULL` failed in one of the 3 ubuntu checks. I couldn't figure out why, any suggestions on how to reproduce the error? I tried to use the same R version `4.0.5` but the unit tests are all passed in my end.

Reply:
Hm, that's tricky.
You could try the code stated in https://pharmaverse.github.io/admiraldev/devel/articles/pr_review_guidance.html#check-fails-only-on-one-version to reproduce the error.

Reply:
Thanks, @bundfussr for the document. I was able to reproduce the error and locate the issue (`tidyr::crossing`), but I'm not sure how to fix this. 
@cicdguy @bms63 would you help check?
below is the issue example from testing environment based on [Check Fails Only on One Version](https://pharmaverse.github.io/admiraldev/devel/articles/pr_review_guidance.html#check-fails-only-on-one-version)
```r
> library(dplyr)
> ids <- tibble::tibble(x=1)[, 0] 
> data_expected <- tibble::tribble(
+   ~PARAMCD, ~AVISITN, ~AVISIT,
+   "a", 1, "WEEK 1",
+   "a", 2, "WEEK 2",
+   "b", 1, "WEEK 1",
+   "b", 2, "WEEK 2"
+ )
> ids %>%
+   tidyr::crossing(data_expected)
# A tibble: 0 x 3
# … with 3 variables: PARAMCD <chr>, AVISITN <dbl>, AVISIT <chr>
```

```r
> sessionInfo()
R version 4.0.5 (2021-03-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/atlas/libblas.so.3.10.3
LAPACK: /usr/lib/x86_64-linux-gnu/atlas/liblapack.so.3.10.3

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8        LC_COLLATE=C.UTF-8    
 [5] LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8    LC_PAPER=C.UTF-8       LC_NAME=C             
 [9] LC_ADDRESS=C           LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_1.0.5

loaded via a namespace (and not attached):
 [1] pillar_1.5.1      compiler_4.0.5    highr_0.8         remotes_2.2.0     tools_4.0.5       digest_0.6.27    
 [7] evaluate_0.14     lifecycle_1.0.0   tibble_3.1.0      pkgconfig_2.0.3   rlang_0.4.10      reprex_1.0.0     
[13] cli_2.3.1         rstudioapi_0.14   curl_4.3          yaml_2.2.1        xfun_0.22         withr_2.4.1      
[19] styler_1.4.1      knitr_1.31        generics_0.1.0    fs_1.5.0          vctrs_0.3.7       tidyselect_1.1.0 
[25] glue_1.4.2        R6_2.5.0          processx_3.5.0    fansi_0.4.2       rmarkdown_2.7     tidyr_1.1.3      
[31] purrr_0.3.4       callr_3.6.0       clipr_0.7.1       magrittr_2.0.1    ps_1.6.0          backports_1.2.1  
[37] ellipsis_0.3.1    htmltools_0.5.1.1 assertthat_0.2.1  utf8_1.2.1        crayon_1.4.1 
```

Same environment but using `renv` produces the right results -
```r
> library(dplyr)
> ids <- tibble::tibble(x=1)[, 0] 
> data_expected <- tibble::tribble(
  +   ~PARAMCD, ~AVISITN, ~AVISIT,
  +   "a", 1, "WEEK 1",
  +   "a", 2, "WEEK 2",
  +   "b", 1, "WEEK 1",
  +   "b", 2, "WEEK 2"
  + )
> ids %>%
  +   tidyr::crossing(data_expected)
# A tibble: 4 x 3
PARAMCD AVISITN AVISIT
<chr>     <dbl> <chr> 
  1 a             1 WEEK 1
2 a             2 WEEK 2
3 b             1 WEEK 1
4 b             2 WEEK 2
```
```r
> sessionInfo()
R version 4.0.5 (2021-03-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/atlas/libblas.so.3.10.3
LAPACK: /usr/lib/x86_64-linux-gnu/atlas/liblapack.so.3.10.3

locale:
[1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8        LC_COLLATE=C.UTF-8    
[5] LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8    LC_PAPER=C.UTF-8       LC_NAME=C             
[9] LC_ADDRESS=C           LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

attached base packages:
  [1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
  [1] dplyr_1.0.5

loaded via a namespace (and not attached):
[1] tidyr_1.1.3      fansi_0.4.2      crayon_1.4.1     utf8_1.2.1       R6_2.5.0         lifecycle_1.0.3 
[7] magrittr_2.0.3   pillar_1.5.1     rlang_1.0.6      cli_3.4.1        renv_0.16.0      rstudioapi_0.13 
[13] vctrs_0.5.0      generics_0.1.0   ellipsis_0.3.2   tools_4.0.5      glue_1.6.2       purrr_0.3.4     
[19] compiler_4.0.5   pkgconfig_2.0.3  tidyselect_1.1.0 tibble_3.1.0    
> 
```

Reply:
I would check the number of columns and call `crossing()` only if there is at least one column. Once we move to tidyr 1.2.0 we can remove the work-around.